### DISCUSSION TITLE
Can we retrieve transformed vectors from tokens (or words) as sparse embeddings?

### DISCUSSION THREAD

**Discussion by lspataroG** [xhluca/bm25s#69] [Q&A] [answered]
Hello @xhluca,
First of all thank you so much for releasing this great library!
I have a question, is it currently supported to retrieve transformed vectors from tokens (or words)?
I am thinking in the style of Sklearn TfidfVectorizer transform. The reason I am asking this is that I would like to use the feature vectors directly. In particular having the functionality to infer sparse vectors from new queries/documents.

Thank you very much!

**Comment by xhluca** [comment, marked as answer]
Thank you for the kind words and glad to hear you enjoy the library!

Right now, it is possible to get a match score (float) between a query and a document; for D documents, that becomes a D-dimension numpy vector. To do this, simply using the `retriever.get_scores` function!

Here's the link: https://github.com/xhluca/bm25s/blob/e1b39e5c0375d20595d5163dc63b84f0a0442ee8/bm25s/__init__.py#L502-L514

I apologize for the lack of docstrings - i haven't had chance to add that as i have been focused on documenting retrieve instead.

If you want the feature vector of a given document wrt to all words in the vocab, that's a bit tricky because bm25s doesn't actually store the sparse matrix, but rather then index pointer (`indptr.csc.index.npy`), indices (`indices.csc.index.npy`) and flat data array (`data.csc.index.npy`), which are stored in a dictionary attributed `retriever.scores`. You can see how it is saved here:

https://github.com/xhluca/bm25s/blob/e1b39e5c0375d20595d5163dc63b84f0a0442ee8/bm25s/__init__.py#L747-L796

Fortunately, you can reconstruct a [scipy csc matrix](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csc_matrix.html) by simply passing this:

```python
# i might have inverted M and N, worth double checking here
M = "<num documents>"
N = "<vocab size>"

doc_matrix = scipy.sparse.csc_matrix((data, indices, indptr), shape=(M, N))
```

Now, you can access the vector for a document wrt all words by selecting your desired column.

**Comment by lspataroG** [comment]
@xhluca Thank you very much! So the second part of the answer is what I was after, I was able to get the feature vector of a given document with respect to the words that it contains. However I am trying to understand whether it is possible to compute that for a new document. The answer above is similar to running a sklearn TfidfVectorizer.fit_transform, now I need to be able to inference on new documents (ie. transform). Is this currently achievable? Thanks!

**Reply by xhluca** [reply]
Unfortunately it's not possible to add new documents. There's relevant discussions on the topic that delves into why it's not possible.

**Reply by lspataroG** [reply]
ok!, thank you :)

**Reply by xhluca** [reply]
Here's one discussion: https://github.com/xhluca/bm25s/discussions/25
